core/src/main/java/com/google/cloud/sql/core/SqlAdminApiFetcher.java

```diff
@@ -318,6 +319,40 @@ private Certificate fetchEphemeralCertificate(
     return ephemeralCertificate;
   }
 
+  /**
+   * refreshWithRetry attempts to refresh the credentials 3 times, waiting 3-6 seconds between
+   * attempts.
+   *
+   * @param credentials the credentials to refresh
+   * @throws IOException when the credentials.refresh() has failed 3 times
+   */
+  private void refreshWithRetry(OAuth2Credentials credentials) throws IOException {
```

Comment:
Should we be adding a test for this? Something like a CredentialRefresher maybe?

Reply:
Good idea. I'll add one.

Reply:
Test added. I refactored the retry logic into its own implementation of Callable called RetryingCallable so that it could be tested.